FabNESO/tasks.py

```diff
@@ -252,3 +267,199 @@ def neso_ensemble(
+        solver,
+        conditions_file_name,
+        mesh_file_name,
+        processes,
+        nodes,
+        cpus_per_process,
+        wall_time,
+    )
+
+    config_dict["neso_args"] = neso_args
+
+    # Hard coded for the two_stream config. Ideally this would be factored out
+    parameters_to_scan = {
+        "particle_initial_velocity": (0.0, 2.0),
+        "particle_charge_density": (20.0, 200.0),
+        "particle_number_density": (20.0, 200.0),
+    }
+
+    config_dict["parameters_to_scan"] = parameters_to_scan
+
+    # Set up the bounds information required by PyVBMC
+    bounds = list(zip(*parameters_to_scan.values(), strict=True))
+    lower_bounds = np.array(bounds[0])
+    upper_bounds = np.array(bounds[1])
+
+    plausible_lower_bounds = lower_bounds + (upper_bounds - lower_bounds) / 4
+    plausible_upper_bounds = upper_bounds - (upper_bounds - lower_bounds) / 4
+
+    # Choose a random starting position
+    rng = np.random.default_rng()
+    theta_0 = rng.uniform(plausible_lower_bounds, plausible_upper_bounds)
+
+    # Run a nominal version of the model with central values to find
+    # the `true' field measurements. Ideally this would be measured data
+    config_dict["initial_run"] = run_instance_return_field(config_dict)
+
+    # Get a timestamp to label the vbmc logfile
+    timestamp = time.strftime("%Y%m%d-%H%M%S")
+    logfile_name = f"vbmc_log_{timestamp}.log"
+    logfile_path = Path(output_directory) / logfile_name
+
+    # Additional options that we will pass to VBMC
+    options = {
+        # 50 * D + 2 is the default, but leave this here to allow
+        # smaller runs for debugging
+        "max_fun_evals": 2 * (len(theta_0) + 2),
+        # Can't see a way of doing this and saving the plots
+        # without it creating a new xwindow?
+        "plot": False,
+        # Run with a log file
+        "log_file_name": logfile_path,
+    }
+
+    # Make an instance of VBMC
+    vbmc = pyvbmc.VBMC(
+        lambda theta: log_density(theta, config_dict),
+        theta_0,
+        lower_bounds,
+        upper_bounds,
+        plausible_lower_bounds,
+        plausible_upper_bounds,
+        options=options,
+    )
+
+    # Run the vbmc instance
+    vp, results = vbmc.optimize()
+
+    # Print the final output of vbmc to the logfile
+    with logfile_path.open("a+") as logfile:
+        logfile.write(pyvbmc.formatting.format_dict(results))
+
+    # Save the final output posterior
+    vbmc.vp.save(output_directory / f"final_posterior_{timestamp}.pkl")
+
+
+def run_instance_return_field(
+    config_dict: dict[str, Any],
+) -> dict[str, np.ndarray]:
+    """Run a single instance of the NESO solver and return the observed_field."""
+    # If we're running remotely, tell the submission to wait until done
+    if "archer2" in fab.env.remote:
+        fab.update_environment(
+            {
+                "job_dispatch": "cd /work/$project/$project/$username ; sbatch --wait",
+            }
+        )
```

Comment:
This is how I'm currently forcing the code to wait after `sbatch` submission. `job_dispatch` is an environment variable that can be edited in the `machines_FabNESO_user.yml` file, and I toyed with just adding it to the example file there, but this feels like a more safe way to force the wait, given the current proposal that we just want it to work on Archer2 for now.

I also tested using the `wait_complete()` method built in to `FabSIM`, and that works perfectly fine (and would be platform independent), but it does have much larger overheard when running short jobs.

Reply:
This seems a good solution for now. Agree that setting as part of task rather than having user need to override in the YAML file is preferable from a robustness perspective and in still allowing asynchronous dispatch of jobs on ARCHER2 in other tasks. Good to also know the `wait_complete` approach is viable - we could possible look at putting that as a fallback option when not using a Slurm based scheduler in future.